MSAL/test/integration/native_auth/end_to_end/sign_up/MSALNativeAuthSignUpUsernameEndToEndTests.swift

```diff
@@ -27,39 +27,37 @@ import MSAL
 
 final class MSALNativeAuthSignUpUsernameEndToEndTests: MSALNativeAuthEndToEndBaseTestCase {
 
-    private let usernameOTP = ProcessInfo.processInfo.environment["existingOTPUserEmail"] ?? "<existingOTPUserEmail not set>"
     private let attributes = ["age": 40]
 
-    override func setUpWithError() throws {
-        try super.setUpWithError()
-        try XCTSkipIf(!usingMockAPI)
-    }
-
     // Hero Scenario 1.1.1. Sign up – with Email Verification (Email & Email OTP)
     func test_signUpWithCode_withEmailVerification_succeeds() async throws {
+        guard let sut = initialisePublicClientApplication(useEmailPasswordClientId: false) else {
+            XCTFail("Missing information")
+            return
+        }
         let codeRequiredExp = expectation(description: "code required")
         let signUpStartDelegate = SignUpStartDelegateSpy(expectation: codeRequiredExp)
-
-        if usingMockAPI {
-            try await mockResponse(.signUpStartSuccess, endpoint: .signUpStart)
-            try await mockResponse(.challengeTypeOOB, endpoint: .signUpChallenge)
-        }
+        let usernameOTP = generateSignUpRandomEmail()
```

Comment:
From what I've tested this particular test, I'm getting ~1 failure out of 10 tries. If we multiply this for every test, the possibility of getting random failures is a bit high.

As we've talked this is due to OTP codes getting stuck on the Microsoft queue.

Right now if a test fails, we try again (up to 3 times), but this generates another email. It's the equivalent to a user that can't get the OTP trying to sign up again with a different email.

Do you know if we can simulate a user clicking on the "resend code" button instead of signing up with a new email?

I'm not sure how this works on the backend side, but (as a user) I've always thought that when I click on resend code it "forces" the backend to send it.

Reply:
It can be, but we need to double check this with the API team before to make this assumption. I personally would prefer to rerun the test for the following reason:
- it doesn't affect the test code. If we ask to the SDK to resend the code, we're interacting with the SDK under test in a way that it is not part of the actual test.
- using the test config is pretty easy to setup the retry logic and we don't need to add new code
- each time we rerun the tests, a new email is created, making it highly unlikely that a new email won't be sent to this address.

Reply:
> it doesn't affect the test code. If we ask to the SDK to resend the code, we're interacting with the SDK under test in a way that it is not part of the actual test.

This is a good point, and it's very valid, but I'm of the opinion that an end to end test should handle a full flow and simulate a real world situation.

> using the test config is pretty easy to setup the retry logic and we don't need to add new code

Another good point, I agree that my suggestion will increase the complexity.

> each time we rerun the tests, a new email is created, making it highly unlikely that a new email won't be sent to this address.

I don't agree with this, if the email is delayed once, I don't understand why signing up with a different email would make the other email be sent on time.

Anyway, all of this is based on the assumption that the API forces the "pending emails" to be sent if a user clicks on "resend code". I would ask the API team first, and if we see the complexity increases too much, we can first monitor the current solution for some time. If we see we have multiple failures for this reason, we can implement it then.

Reply:
> but I'm of the opinion that an end to end test should handle a full flow and simulate a real world situation.
The test scenario should be designed for consistent replication, ensuring it evaluates the same behavior each time. Introducing additional steps for certain situations go against this requirement.
> don't agree with this, if the email is delayed once, I don't understand why signing up with a different email would make the other email be sent on time.
The issue isn't with the entire email OTP service being non-functional; it's specific to a particular email address where the process seems to halt. From what I've observed, initiating the signup using an alternative email address resolves the problem.
If we notice that we have many test emails don't receive the email OTP on time, it might be good to consult eSTS for further info, as this could potentially impact our customers.